### PR TITLE
#5438: Add kokkos configuration to benchmark metadata

### PR DESCRIPTION
**Pull request by thearusable**
Related to #5348 

This PR is adding the possibility to add kokkos configuration data to benchmark context. This way it can be included in json output file after running benchmarks.

<details>
  <summary>Example JSON</summary>

```json
{
  "context": {
    "date": "2022-08-24T12:05:55+02:00",
    "host_name": "macbook-pro.home",
    "executable": "/Users/arus/NGA/Kokkos/build/kokkos/core/perf_test/KokkosCore_PerformanceTest_Benchmark",
    "num_cpus": 8,
    "mhz_per_cpu": 24,
    "cpu_scaling_enabled": false,
    "caches": [
      {
        "type": "Data",
        "level": 1,
        "size": 65536,
        "num_sharing": 0
      },
      {
        "type": "Instruction",
        "level": 1,
        "size": 131072,
        "num_sharing": 0
      },
      {
        "type": "Unified",
        "level": 2,
        "size": 4194304,
        "num_sharing": 1
      }
    ],
    "load_avg": [2.00293,1.97412,1.46094],
    "library_build_type": "release",
    "Default Device": "N6Kokkos6OpenMPE",
    "KOKKOS_COMPILER_APPLECC": "1",
    "KOKKOS_COMPILER_GNU": "1210",
    "KOKKOS_ENABLE_ASM": "no",
    "KOKKOS_ENABLE_CXX17": "yes",
    "KOKKOS_ENABLE_CXX20": "no",
    "KOKKOS_ENABLE_CXX23": "no",
    "KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK": "yes",
    "KOKKOS_ENABLE_GNU_ATOMICS": "no",
    "KOKKOS_ENABLE_HBWSPACE": "no",
    "KOKKOS_ENABLE_HWLOC": "no",
    "KOKKOS_ENABLE_INTEL_ATOMICS": "no",
    "KOKKOS_ENABLE_INTEL_MM_ALLOC": "no",
    "KOKKOS_ENABLE_LIBDL": "yes",
    "KOKKOS_ENABLE_LIBRT": "no",
    "KOKKOS_ENABLE_OPENMP": "yes",
    "KOKKOS_ENABLE_OPENMP_ATOMICS": "no",
    "KOKKOS_ENABLE_PRAGMA_IVDEP": "no",
    "KOKKOS_ENABLE_PRAGMA_LOOPCOUNT": "no",
    "KOKKOS_ENABLE_PRAGMA_UNROLL": "no",
    "KOKKOS_ENABLE_PRAGMA_VECTOR": "no",
    "KOKKOS_ENABLE_WINDOWS_ATOMICS": "no",
    "Kokkos": "OpenMP thread_pool_topology[ 1 x 8 x 1 ]",
    "Kokkos Version": "3.7.99"
  },
  "benchmarks": [
    {
      "name": "ViewDeepCopy_Rank8<Kokkos::LayoutLeft, Kokkos::LayoutLeft>/N:10/R:1/manual_time",
      "family_index": 0,
      "per_family_instance_index": 0,
      "run_name": "ViewDeepCopy_Rank8<Kokkos::LayoutLeft, Kokkos::LayoutLeft>/N:10/R:1/manual_time",
      "run_type": "iteration",
      "repetitions": 1,
      "repetition_index": 0,
      "threads": 1,
      "iterations": 9,
      "real_time": 5.6612000000000010e-02,
      "cpu_time": 5.0412777777777779e-02,
      "time_unit": "s",
      "GB/s": 2.7338068860600760e+01,
      "MB": 7.6293945312500000e+02
    },
    (...)
  ]
}
```

</details>

### CODE REVIEW

**Comment by cz4rs** [comment]
This is interesting (and probably incorrect):
```json
    "KOKKOS_COMPILER_APPLECC": "1",
    "KOKKOS_COMPILER_GNU": "1210",
```
I expect NVCC and NVHPC could report `GNU` apart from their own id as well.

We should probably try to make a single `"compiler": "compiler-version"` entry out of this.
Also, this could possibly be a bug in compiler macros implementation in `Kokkos_Macros.hpp`.

**Comment by thearusable** [comment]
> This is interesting (and probably incorrect):
> 
> ```json
>     "KOKKOS_COMPILER_APPLECC": "1",
>     "KOKKOS_COMPILER_GNU": "1210",
> ```
> 
> I expect NVCC and NVHPC could report `GNU` apart from their own id as well.
> 
> We should probably try to make a single `"compiler": "compiler-with-version"` entry out of this. Also, this could be considered a bug in compiler macros implementation in `Kokkos_Macros.hpp`.

Looks like those values are correct. Those are my gcc definition related to apple: 
```
gcc-12 -dM -E - < /dev/null | egrep -i 'os_|mac|apple'
#define __APPLE_CC__ 1
#define __APPLE__ 1
#define __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ 120000
#define __MACH__ 1
```

**Comment by crtrott** [comment]
I agree this is correct. Lets leave it that way. For stuff like nvcc I also want both showing up.

**Review by crtrott — APPROVED** [review]

File: core/perf_test/Benchmark_Context.hpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+void add_kokkos_configuration(bool verbose = false) {
+  std::ostringstream msg;
+  Kokkos::print_configuration(msg, verbose);
+
+  std::stringstream ss{msg.str()};
+  for (std::string line; std::getline(ss, line, '\n');) {
+    auto found = line.find_first_of(':');
+    if (found != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/perf_test/Benchmark_Context.hpp:71]: Note that this is brittle in the sense that will break if any of the backend does not follow the format or if we change it.
You rely on
CATEGORY-NAME(COLON)NO-TRAILING-WHITESPACE
FOO(NO-TRAILING-WHITESPACE-SPACE)(COLON)BAR(NO-TRAILING-WHITESPACE)

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/perf_test/Benchmark_Context.hpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      auto val = remove_unwanted_prefix(line.substr(found + 1));
+      if (val.length()) {
+        benchmark::AddCustomContext(
+            remove_unwanted_prefix(line.substr(0, found)), val);
+      }
+    }
+  }
+}
+
+void add_benchmark_context(bool verbose = false) {
```
[the remark this answers is not in the document]
Comment (thearusable) [reply core/perf_test/Benchmark_Context.hpp:81]: We need that value in `add_kokkos_configuration` function which calls the `Kokkos::print_configuration` to get the more verbose output. Alternatively we can hardcode it to always return verbose output. 

`add_benchmark_context` is the main function which will be called before running benchmark.

**Comment by masterleinad** [comment]
Might interfere with #5449 which introduces some more whitespace for readability.

File: core/perf_test/Benchmark_Context.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/perf_test/Benchmark_Context.hpp:81]: I am not super convinced that it is useful while it adds complexity but I won't block

File: core/perf_test/Benchmark_Context.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/perf_test/Benchmark_Context.hpp:71]: Add a comment in the code

[the remark this answers is not in the document]
Comment (thearusable) [reply core/perf_test/Benchmark_Context.hpp:80]: I added comments in the code

**Review by cz4rs — APPROVED** [review]
Looks good to me, I'll just rebase if this one goes in first.

File: core/perf_test/BenchmarkMain.cpp [hunk 3]
```diff
@@ -44,4 +44,18 @@
...
 
-BENCHMARK_MAIN();
+#include <Benchmark_Context.hpp>
+#include <Kokkos_Core.hpp>
+
+int main(int argc, char** argv) {
+  Kokkos::initialize(argc, argv);
+  benchmark::Initialize(&argc, argv);
+  benchmark::SetDefaultTimeUnit(benchmark::kSecond);
+  Test::add_benchmark_context(true);
```
Comment (dalg24) [inline core/perf_test/BenchmarkMain.cpp:54]:
```suggestion
  KokkosBenchmark::add_benchmark_context(true);
```
In my previous comment I implicitly meant to update this as well

Comment (thearusable) [reply core/perf_test/BenchmarkMain.cpp:54]: Ah, yes I forgot that there is also that namespace used. Thanks!

**Pull request merged**